### PR TITLE
New tutorial using interactive API on the brainMRI segmentation dataset

### PR DESCRIPTION
**Pull request by saransh09**
* A new segmentation tutorial using the latest version of Interactive API of OpenFL
* Using the Medical Decathlon BrainMRI dataset
* Tried to make the tutorial more informative and dumbed it down a little to make it smoother to use for general devolopers
* Till now most of the tutorials were using a dataset which was very small (in the order of 50-60MBs)
* This dataset used >10GB, will be good to demonstrate how OpenFL scales and can be used on larger datasets
* has a custom shard_descriptor which is quite different from the KVASIR example (as the data is organised very differently)
* Using CSVs to keep a track of the data

### CODE REVIEW

**Comment by openfl-helper** [comment]
Can one of the admins verify this patch?

File: openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/prepare_dataset.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        with open(valid_data_csv,'a') as f:
+            f.write(f'{img_name},{img_slice_name},{msk_slice_name}\n')
+
+# Copy the dataset to the respective folders
+os.system('cp -r processed_data envoy_1/')
+os.system('cp -r processed_data envoy_2/')
+os.system('cp processed_data/train_data.csv envoy_1/')
+os.system('cp processed_data/train_data.csv envoy_2/')
+os.system('cp processed_data/valid_data.csv envoy_1/')
+os.system('cp processed_data/valid_data.csv envoy_2/')
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/prepare_dataset.py]: @saransh09 , can we refactor code to add the following at the end of the file?
```python3
if __name__ == '__main__':
    main()
```

**Comment by alexey-khorkin** [comment]
@saransh09 do we really need `envoy_2` folder? Please, check other examples.

File: openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy_1/shard_descriptor.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
+# Copyright (C) 2020-2021 Intel Corporation
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy_1/shard_descriptor.py]: @saransh09 could we add type annotations? Like here - https://github.com/intel/openfl/blob/develop/openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/envoy/mnist_shard_descriptor.py

**Review by alexey-khorkin — CHANGES_REQUESTED** [review]
Need changes

**Comment by saransh09** [comment]
@alexey-khorkin 
I deliberately kept the envoy_2 folder because I wanted the two envoy setup to be clearer. However, I understand that to maintain the homogeneity of the examples. I will remove it. There is one problem here, that the data is present in google drive and there is no way to download it through the link. The two folders (for envoys) were made so that the script would process and copy the data in the respective folders as needed. Else, it would become a little too opaque to run the experiment for multiple envoys

File: openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/prepare_dataset.py (hunk 1 above)
Comment (saransh09) [reply openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/prepare_dataset.py]: > @saransh09 , can we refactor code to add the following at the end of the file?
> 
> ```python
> if __name__ == '__main__':
>     main()
> ```

Yes sure, I will do this, on it.

File: openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy_1/shard_descriptor.py (hunk 2 above)
Comment (saransh09) [reply openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy_1/shard_descriptor.py]: Sure, I am on it.

**Comment by alexey-gruzdev** [comment]
@saransh09 could you please drop a comment with signing CLA - `I have read the CLA Document and I hereby sign the CLA`, otherwise next stages with Jenkins build are skipped.

**Comment by saransh09** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by alexey-gruzdev** [comment]
ok to test

**Comment by alexey-gruzdev** [comment]
Looks like CI is not green, I believe you need first fix - [Linter](http://213.221.44.203/job/Federated-Learning/job/precommit/1496/execution/node/32/log/)

**Comment by saransh09** [comment]
> Looks like CI is not green, I believe you need first fix - [Linter](http://213.221.44.203/job/Federated-Learning/job/precommit/1496/execution/node/32/log/)

Thanks for this, reading the comments and changing everything. I will make all the changes and make another PR in some time

**Comment by alexey-gruzdev** [comment]
> > Looks like CI is not green, I believe you need first fix - [Linter](http://213.221.44.203/job/Federated-Learning/job/precommit/1496/execution/node/32/log/)
> 
> Thanks for this, reading the comments and changing everything. I will make all the changes and make another PR in some time

You don't need to open another PR. Just add fixes on top of your current branch.

**Comment by alexey-khorkin** [comment]
> There is one problem here, that the data is present in google drive and there is no way to download it through the link. 


@saransh09  we use `gdown` to download files from google drive, for example: https://github.com/intel/openfl/blob/7af3fc8f0ac0328e19fb7b407c36525edb1506ae/openfl-tutorials/interactive_api/Tensorflow_Word_Prediction/envoy/shard_descriptor.py#L137

**Comment by saransh09** [comment]
> > There is one problem here, that the data is present in google drive and there is no way to download it through the link.
> 
> @saransh09 we use `gdown` to download files from google drive, for example:
> 
> https://github.com/intel/openfl/blob/7af3fc8f0ac0328e19fb7b407c36525edb1506ae/openfl-tutorials/interactive_api/Tensorflow_Word_Prediction/envoy/shard_descriptor.py#L137

Done. Added it and refactored it into the shard_descriptor to be consistent to the other tutorials. Thanks for the suggestion.

**Review by alexey-khorkin — CHANGES_REQUESTED** [review]
Please, remove data files from PR: 
* `openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/train_data.csv`
* `openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/valid_data.csv`

**Comment by saransh09** [comment]
> Please, remove data files from PR:
> 
> * `openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/train_data.csv`
> * `openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/valid_data.csv`

Done!

**Comment by saransh09** [comment]
The Jenkins pipeline has been running for ~24 hours now. Is there something wrong from my end?

**Comment by itrushkin** [comment]
> The Jenkins pipeline has been running for ~24 hours now. Is there something wrong from my end?

Looks like CI is stuck. There is an issue on Jenkins side. Restarting the build.

**Comment by itrushkin** [comment]
Jenkins please retry a build

File: openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/.gitignore [hunk 3]
```diff
@@ -0,0 +1,12 @@
+Task01_BrainTumour.tar
```
Comment (igor-davidyuk) [inline openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/.gitignore:1]: This file should be removed

File: openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+                    if ((start + cropLen) > imgLen):  # Don't fall off the image
+                        start = (imgLen - cropLen) // 2
+            else:
+                start = 0
+
+            slices.append(slice(start, start + cropLen))
+
+        return img[tuple(slices)], msk[tuple(slices)]
+
+    def augment_data(self, img, msk) -> Tuple[Any, Any]:
```
Comment (igor-davidyuk) [inline openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py:110]: Augmenting data is a part of FL experiment. Shard descriptor just provides local data

File: openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class brainMRIShard(ShardDataset):
+    """brainMRI Shard dataset class."""
+
+    def __init__(self, csv_path, rank=1, worldsize=1,
+                 crop_dims=None, dataset_dir='./',
+                 augment=True, device='cpu', set='train') -> None:
```
Comment (igor-davidyuk) [inline openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py:29]:
```suggestion
                 augment=True, set='train') -> None:
```
Managing devices also is no concern for a Shard Descriptor

File: openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+                self.csv = self.csv.loc[(self.rank - 1) * (
+                    len(self.csv) // self.worldsize):, :].reset_index(
+                        drop=True)
+            else:
+                self.csv = self.csv.loc[(self.rank - 1) * (len(
+                    self.csv) // self.worldsize):(self.rank) * (len(
+                        self.csv) // self.worldsize), :].reset_index(drop=True)
+        self.device = device
+        self.augment = augment
+
```
Comment (igor-davidyuk) [inline openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py:48]:
```suggestion


```

File: openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+        img = np.load(img_path)
+        msk = np.load(msk_path)
+        if self.crop_dim is None:
+            self.crop_dim = [img.shape[0], img.shape[1]]
+        img = self.preprocess_img(img)
+        msk = self.preprocess_label(msk)
+        img, msk = self.crop_input(img, msk)
+        if self.augment:
+            img, msk = self.augment_data(img, msk)
+        img = img[np.newaxis, :, :].copy()
```
Comment (igor-davidyuk) [inline openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py:62]: Augmenting should be moved to experiment's Dataloader part

File: openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+        num_train = int(np.floor(num_images * .80))
+        train_image_list = image_list[:num_train]
+        valid_image_list = image_list[num_train:]
+        # The processed images are saved in ./processed_data/ folder
+        save_folder = './processed_data/'
+        if os.path.exists(save_folder):
+            print("processed_data/ already exists, do you want to proceed? \
+                press y/Y to proceed and n/N to cancel")
+            inp = input()
+            if inp == 'y' or inp == 'Y':
```
Comment (mansishr) [inline openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py:226]: Processing images and masks shouldn't be under download_data function.

File: openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/workspace/layers.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
+# Copyright (C) 2020-2021 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+"""Layers for Unet model."""
+
```
Comment (mansishr) [inline openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/workspace/layers.py:4]: demo images under workspace should be removed

File: openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/workspace/BrainMRI2D_Segmentation_OpenFL_Tutorial.ipynb [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+    "from torch.utils.data import Dataset, DataLoader, SubsetRandomSampler\n",
+    "from torchvision import transforms as tsf\n",
+    "from copy import deepcopy\n",
+    "import torch\n",
+    "import torch.nn as nn\n",
+    "import torch.optim as optim\n",
+    "import torch.nn.functional as F \n",
+    "import tqdm\n",
+    "from openfl.component.aggregation_functions import Median\n",
+    "from layers import soft_dice_coef, soft_dice_loss, DoubleConv, Down, Up"
```
Comment (mansishr) [inline openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/workspace/BrainMRI2D_Segmentation_OpenFL_Tutorial.ipynb:112]: soft_dice_coef and soft_dice_loss are being redefined in the notebook

File: openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/workspace/BrainMRI2D_Segmentation_OpenFL_Tutorial.ipynb [hunk 11]
```diff
@@ -0,0 +1,476 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "120c95ec",
+   "metadata": {},
+   "source": [
+    "### Download the dataset : Medicat Decathlon BrainMRI Segmentation Dataset\n",
+    "\n",
+    "Download the data from : https://goo.gl/QzVZcm download the file Task01_BrainTumour.tar (into the base repository) <p>\n",
+    "Then run the prepare_dataset.py file to process and create the dataset so that it is ready to use. <p>\n",
```
Comment (mansishr) [inline openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/workspace/BrainMRI2D_Segmentation_OpenFL_Tutorial.ipynb:19]: The instructions to download the dataset/setup etc should be a part of README and removed from the notebook.

**Review by mansishr — CHANGES_REQUESTED** [review]

**Comment by theakshaypant** [comment]
@saransh09 can you please resolve the comments and request for review again?

**Comment by psfoley** [comment]
Closing due to deprecation of the Interactive API
